src/consensus/aft/raft.h

```diff
@@ -1369,29 +1398,51 @@ namespace aft
         }
       }
 
-      send_append_entries_response(from, AppendEntriesResponseType::OK);
+      send_append_entries_response_ack(from, r);
+    }
+
+    void send_append_entries_response_ack(
+      ccf::NodeId to, const AppendEntries& ae)
+    {
+      // If we get to here, we have applied up to r.idx in this AppendEntries.
+      // We must only ACK this far, as we know nothing about the agreement of a
+      // suffix we may still hold _after_ r.idx with the leader's log
+      const auto response_idx = ae.idx;
+      send_append_entries_response(
+        to, AppendEntriesResponseType::OK, state->current_view, response_idx);
+    }
```

Comment:
This is the second key fix, the "only ACK up to incoming AE" change. I was unable to reach this path without the rollback_safety fix, but after that this becomes a secondary clear safety violation.